src/internal/clusterstate/v2.10.0/pipelines.go

```diff
@@ -0,0 +1,239 @@
+package v2_10_0
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/migrations"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/pachyderm/pachyderm/v2/src/pps"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+var CreateUniqueIndex = `
+  CREATE UNIQUE INDEX pip_version_idx ON collections.pipelines(idx_name,idx_version);
+`
+
+// find all pipelines with duplicate versions, and return all of the pipeline versions for each of those pipelines
+var duplicatePipelinesQuery = `
+  SELECT key, proto
+  FROM collections.pipelines
+  WHERE idx_name
+    IN (
+      SELECT idx_name
+      FROM collections.pipelines
+      GROUP BY idx_name, idx_version
+      HAVING COUNT(key) > 1
+    )
+  ORDER BY idx_name, createdat;
+`
+
+func DeduplicatePipelineVersions(ctx context.Context, env migrations.Env) error {
+	pipUpdates, pipVersionChanges, err := collectPipelineUpdates(ctx, env.Tx)
+	if err != nil {
+		return err
+	}
+	log.Info(ctx, "detected updates for pipeline versions", zap.Int("updates_count", len(pipUpdates)))
+	if err := UpdatePipelineRows(ctx, env.Tx, pipUpdates); err != nil {
+		return err
+	}
+	if err := UpdateJobPipelineVersions(ctx, env.Tx, pipVersionChanges); err != nil {
+		return err
+	}
+	if _, err := env.Tx.ExecContext(ctx, CreateUniqueIndex); err != nil {
+		return errors.Wrap(err, "create unique index pip_version_idx")
+	}
+	return nil
+}
+
+func collectPipelineUpdates(ctx context.Context, tx *pachsql.Tx) (rowUpdates []*PipUpdateRow,
+	pipelineVersionChanges map[string]map[uint64]uint64,
+	retErr error) {
+	rr, err := tx.QueryxContext(ctx, duplicatePipelinesQuery)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "query duplicates in collections.pipelines")
+	}
+	defer rr.Close()
+	pi := &pps.PipelineInfo{}
+	pipLatestVersion := make(map[string]uint64)
+	pipVersionChanges := make(map[string]map[uint64]uint64)
+	var updates []*PipUpdateRow
+	for rr.Next() {
+		var row pipDBRow
+		if err := rr.Err(); err != nil {
+			return nil, nil, errors.Wrap(err, "row error")
+		}
+		if err := rr.StructScan(&row); err != nil {
+			return nil, nil, errors.Wrap(err, "scan pipeline row")
+		}
+		if err := proto.Unmarshal(row.Proto, pi); err != nil {
+			return nil, nil, errors.Wrapf(err, "unmarshal proto")
+		}
+		lastChange, ok := pipLatestVersion[pi.Pipeline.Name]
```

Comment:
I'm not actually sure that the rows are being visited in version order?  If you visited 3, 2, 1, they'd get renumbered to 1, 2, 3?  That's not what we want.

Reply:
`duplicatePipelinesQuery` orders results by createdat